### PR TITLE
ROX-18428: report secret data to fleet-manager

### PR DESCRIPTION
**Pull request by johannes94**
## Description

This PR adds the logic to report K8s secrets created by FS sync or the RHACS operator back to store them in fleet-manager (namely `central-tls` and `central-db-password`). The secrets are encrypted using AWS KMS to ensure they are stored securely in FMs database and in transit. We need to create the KMS keys before merging this into main and deploying it to the stage environment. Because of that the PR: https://github.com/stackrox/acs-fleet-manager-aws-config/pull/117 has to be merged first.

This is part of the Epic https://issues.redhat.com/browse/ROX-17060, the implementation to restore secrets if they are missing in the cluster will be done in another PR.

In this PR we implement following parts of the overall solution:
- Getting the secret from cluster
- Encrypting them
- Report from fleetshard-sync to fleet-manager
- Store them in fleet-manager DB
- Report if secrets are stored or not

## Checklist (Definition of Done)

- [x] Unit and integration tests added
- [x] Added test description under `Test manual`
- [x] Documentation added if necessary (i.e. changes to dev setup, test execution, ...)
- [x] CI and all relevant tests are passing
- [x] Add the ticket number to the PR title if available, i.e. `ROX-12345: ...`
- [x] Discussed security and business related topics privately. Will move any security and business related topics that arise to private communication channel.
- ~~[ ] Add secret to app-interface Vault or Secrets Manager if necessary~~
- ~~[ ] RDS changes were e2e tested [manually](../docs/development/howto-e2e-test-rds.md)~~
- [x] Check AWS limits are reasonable for changes provisioning new resources

## Test manual

```
# Create an OSD cluster using the rhacs-managed-service-dev account
# Provision enough resources for an internal PostgreSQL DB at this point at least 1 node with 8 CPUs

make binary
make db/teardown db/setup db/migrate

# Login to OSD cluster and install RHACS Operator
# Configure AWS CLIE to point to dev
# Create a fleet-manager cluster-config for the OSD cluster by editing /dev/config/dataplane-cluster-configuration-infractl-osd.yaml

# Start fleet-manager and fleetshard-sync in different shells
# Configure fleetshard-sync
export AWS_REGION=us-east-1
export SECRET_ENCYRPTION_TYPE=kms
export OCM_TOKEN=$(ocm token --refresh)
export AUTH_TYPE="ocm"
export SECRET_ENCRYPTION_KEY_ID=<dev secret encryption key>
export CLUSTER_ID=<your-cluster-id>
./fleetshard-sync

# Run fleet-manager
./fleet-manager serve --dataplane-cluster-config-file "./dev/config/dataplane-cluster-configuration-infractl-osd.yaml"

# Create a central and make sure it get's to ready state without errors
./scripts/create-central.sh

# Make sure secrets are populated in the database
make db/psql
select id, status, secrets from central_requests;



```

```
# To run tests locally run:
make db/teardown db/setup db/migrate
make ocm/setup OCM_OFFLINE_TOKEN=<ocm-offline-token> OCM_ENV=development
make verify lint binary test test/integration
```

### CODE REVIEW

File: openapi/fleet-manager-private.yaml [hunk 1]
```diff
@@ -234,6 +234,10 @@ components:
                       type: string
                 deletionTimestamp:
                   type: string
+                secretsStored:
```
Comment (johannes94) [inline openapi/fleet-manager-private.yaml:237]: I only report secrets as a list of secret names, because the secrets data itself is at ~10KB per tenant. To save network traffic and costs I'll implement lazy loading of secret data in the follow up PR for secret restore in fleetshard-sync.

File: fleetshard/pkg/k8s/secret.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	centralDBPasswordSecretName = "central-db-password" // pragma: allowlist secret
+)
+
+var secretsToWatch = []string{
+	centralTLSSecretName,
+	centralDBPasswordSecretName,
+}
+
+// SecretService is responsible for reading and writing secrets
+// This servise is specific to ACS Managed Services and provides methods work on specific secrets
```
Comment (vladbologa) [inline fleetshard/pkg/k8s/secret.go:25]:
```suggestion
// SecretService is responsible with reading specified secrets from a Kubernetes namespace
```

There's no writing as of yet, I guess you plan to add it in the future, but for now it's misleading. Also the name SecretService sounds a bit funny, maybe we can come up with another name?

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 3]
```diff
@@ -739,9 +744,90 @@ func (r *CentralReconciler) collectReconciliationStatus(ctx context.Context, rem
 		}
 	}
 
+	// Only report secrets if central is ready to ensure we're not trying to get secrets before they are created
+	// also only report secrets once to don't overwrite initial secrets with possibly corrupted secrets
+	// from the cluster state.
```
Comment (vladbologa) [inline fleetshard/pkg/central/reconciler/reconciler.go:749]:
```suggestion
	// Only report secrets if Central is ready, to ensure we're not trying to get secrets before they are created
	// Also, only report secrets once, to not overwrite initial secrets with possibly corrupted secrets
	// from the cluster state.
```

File: internal/dinosaur/pkg/services/data_plane_dinosaur.go [hunk 4]
```diff
@@ -75,7 +75,7 @@ func (d *dataPlaneCentralService) UpdateDataPlaneCentralService(ctx context.Cont
 		case statusReady:
 			// Only store the routes (and create them) when the Dinosaurs are ready, as by the time they are ready,
 			// the routes should definitely be there.
-			e = d.persistCentralRoutes(dinosaur, ks, cluster)
+			e = d.persistCentralValues(dinosaur, ks, cluster)
```
Comment (vladbologa) [inline internal/dinosaur/pkg/services/data_plane_dinosaur.go:78]: Please update the comment above.

**Review by vladbologa — CHANGES_REQUESTED** [review]
Looks good, I had a few comments here and there.

File: internal/dinosaur/pkg/services/data_plane_dinosaur.go (hunk 4 above)
Comment (johannes94) [reply internal/dinosaur/pkg/services/data_plane_dinosaur.go:78]: Done.

File: fleetshard/pkg/k8s/secret.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+)
+
+const (
+	centralTLSSecretName        = "central-tls"         // pragma: allowlist secret
+	centralDBPasswordSecretName = "central-db-password" // pragma: allowlist secret
+)
+
+var secretsToWatch = []string{
+	centralTLSSecretName,
+	centralDBPasswordSecretName,
```
[the remark this answers is not in the document]
Comment (johannes94) [reply fleetshard/pkg/k8s/secret.go:21]: While I see that this solution is simple to implement introducing edge cases for specific secrets might be harder to maintain. Consider we now have:
- central-tls
- db-password

We would need to document as SOPS and test (perhaps automatically) 2 different approaches to restore/regenerate those secrets. Lets say we add a KMS key, now we have to ask ourselfes, can we easily regenerate it, should we backup it? Is there a potential 3rd solution to recover. We would need to go through SOPs and add it everywhere and need to write a new one if we implement something different.

That's why I'd like to handle both generated secrets the same for now. WDYTH? IMO we can also defer the decision to the point where we implement db password rotation.

[the remark this answers is not in the document]
Comment (vladbologa) [reply fleetshard/pkg/k8s/secret.go:20]: I'm ok with deferring the decision until we implement db password rotation.

But to continue the discussion a bit. When dealing with secrets created and managed by `fleetshard-sync` (currently only the db-password falls into this category afaik), if they are missing, fleetshard-sync could and should be capable of regenerating them and recover on its own. So we wouldn't even need to mention them in SOPs.

File: internal/dinosaur/pkg/migrations/20230802000000_add_secrets_field_to_central_request.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		QuotaType                     string     `json:"quota_type"`
+		Routes                        api.JSON   `json:"routes"`
+		Secrets                       api.JSON   `json:"secrets"`
+		RoutesCreated                 bool       `json:"routes_created"`
+		Namespace                     string     `json:"namespace"`
+		RoutesCreationID              string     `json:"routes_creation_id"`
+		DeletionTimestamp             *time.Time `json:"deletionTimestamp"`
+		Central                       api.JSON   `json:"central"`
+		Scanner                       api.JSON   `json:"scanner"`
+		OperatorImage                 string     `json:"operator_image"`
```
[the remark this answers is not in the document]
Comment (vladbologa) [reply internal/dinosaur/pkg/migrations/20230802000000_add_secrets_field_to_central_request.go:50]: Yes, looks like in some migrations we do, in others we don't.

I remember first seeing a comment about this here: https://github.com/stackrox/acs-fleet-manager/pull/904#discussion_r1147656127

After that I also brought it up on other occasions. What I don't know is, does it bring any benefits to have the full model? If not,  I think having less code is generally better.

[the remark this answers is not in the document]
Comment (johannes94) [reply internal/dinosaur/pkg/migrations/20230802000000_add_secrets_field_to_central_request.go:50]: I've reduced it to only the secrets field

**Review by vladbologa — APPROVED** [review]

File: fleetshard/pkg/central/reconciler/reconciler.go [hunk 7]
```diff
@@ -104,6 +107,8 @@ type CentralReconciler struct {
 	useRoutes        bool
 	Resources        bool
 	routeService     *k8s.RouteService
+	secretService    *k8s.SecretBackup
```
Comment (vladbologa) [inline fleetshard/pkg/central/reconciler/reconciler.go:110]:
```suggestion
	secretBackup    *k8s.SecretBackup
```

Comment (johannes94) [reply fleetshard/pkg/central/reconciler/reconciler.go]: Done.
